python/lsst/ts/tunablelaser/csc.py

```diff
@@ -0,0 +1,261 @@
+
+        Parameters
+        ----------
+        data
+        """
+        self.assert_enabled("stopPropagateLaser")
+        self.assert_substate([TunableLaser.LaserDetailedState.PROPAGATING], "stopPropagateLaser")
+        try:
+            self.model.stop()
+        except Exception as e:
+            self.evt_errorCode(
+                errorCode=TunableLaser.LaserErrorCode.general_error,
+                errorReport=e.msg,
+                traceback=traceback.format_exc()
+            )
+        self.detailed_state = TunableLaser.LaserDetailedState.ENABLEDSTATE
+
+    async def do_clearFaultState(self, data):
+        """Clear the fault state of the laser by turning the power register
+        off.
+
+        Parameters
+        ----------
+        data
+        """
+        self.model.clear_fault()
+
+    @property
+    def detailed_state(self):
+        """Return the current substate of the laser and when it changes
+        publishes an event.
+        """
+        return TunableLaser.LaserDetailedState(self.evt_detailedState.data.detailedState)
+
+    @detailed_state.setter
+    def detailed_state(self, new_sub_state):
+        new_sub_state = TunableLaser.LaserDetailedState(new_sub_state)
+        self.evt_detailedState.set_put(detailedState=new_sub_state)
+
+    async def begin_enable(self, data):
+        """A hook that sets up the laser for propagation so that it
+        is ready to go.
+
+        Parameters
+        ----------
+        data
+        """
+        try:
+            self.model._laser.MaxiOPG.set_configuration(self.model.csc_configuration.optical_configuration)
+            self.model._laser.set_output_energy_level("MAX")
+        except TimeoutError as te:
+            self.fault(code=TunableLaser.LaserErrorCode.timeout_error, report=te.msg)
+
+    async def begin_disable(self, data):
+        """
+
+        Parameters
+        ----------
+        data
+        """
+        if self.model._laser.M_CPU800.power_register_2.register_value == "ON":
+            try:
+                self.model.stop()
+                self.detailed_state = TunableLaser.LaserDetailedState(
+                    TunableLaser.LaserDetailedState.ENABLEDSTATE)
+            except TimeoutError as te:
+                self.fault(code=TunableLaser.LaserErrorCode.timeout_error, report=te.msg)
+        self.model.set_output_energy_level("OFF")
+
+    async def end_start(self, data):
+        try:
+            self.model.connect()
+            self.telemetry_task = asyncio.ensure_future(self.telemetry())
+        except Exception as e:
+            self.evt_errorCode(
+                errorCode=TunableLaser.LaserErrorCode.general_error,
+                errorReport=e.msg,
+                traceback=traceback.format_exc()
+            )
+
+    async def end_standby(self, data):
+        try:
+            if not self.telemetry_task.done():
+                self.telemetry_task.set_result('done')
+            self.model.disconnect()
```

Comment:
I *very strongly* suggest moving this to handle_summary_state. That way it can be run whenever you go out of enabled-or-disabled state. In particular you can disconnect the model when you go into FAULT state.